### ISSUE TITLE
How to incorporate poetry self commands in an environment

### ISSUE THREAD

**Issue by ianepreston** [cachix/devenv#1233]
I need to add the `artifacts-keyring` package into my poetry install to be able to interact with Azure Artifacts Feed. Outside of nix I'd run `poetry self add artifacts-keyring`, which would install the package into poetry's virtual environment (not my project's). Looking at the commands available under poetry for devenv I don't see where I'd specify that. There's an option to specify a poetry package, would I have to do it there somehow by making my own poetry derivation and adding it in?

Thanks

**Comment by sandydoo** [comment]
@ianepreston, I've never used poetry plugins before, but maybe something like this?
One issue is that I can't seem to find `artifacts-keyring` in nixpkgs, so we need to package that first.

Would you be able to provide a sample project for us to test out this integration? Ideally, something self-contained where we can add the plugin and somehow verify that it works.

```nix
{ pkgs, lib, config, inputs, ... }:

{
  languages.python.enable = true;
  languages.python.poetry.enable = true;
  languages.python.poetry.package = pkgs.poetry.withPlugins (
    ps: [
      config.languages.python.poetry.package.pkgs.artifacts-keyring # Not in nixpkgs
    ]
  );
}
```

**Comment by ianepreston** [comment]
@sandydoo - thanks for the quick reply! I'll have to think about a good way to make a self-contained example. Artifacts keyring really only works if you have a DevOps artifacts feed to interact with and I don't have access to a public one.
As an easier start could we figure out how to incorporate [poetryPlugins.poetry-plugin-export](https://search.nixos.org/packages?channel=23.11&show=poetryPlugins.poetry-plugin-export&from=0&size=50&sort=relevance&type=packages&query=poetry)?

That seems like it would be easier to test, and it's already packaged. If we can get that going then I assume the next step would be to use it as a template to build an equivalent `poetryPlugins.artifacts-keyring` package. Does that make sense?

If so I'll dig into the plugin export functionality and stand up a sample project with it.

**Comment by sandydoo** [comment]
> As an easier start could we figure out how to incorporate [poetryPlugins.poetry-plugin-export](https://search.nixos.org/packages?channel=23.11&show=poetryPlugins.poetry-plugin-export&from=0&size=50&sort=relevance&type=packages&query=poetry)?

I think that plugin is even passed through by `withPlugins`:
```nix
  languages.python.poetry.package = pkgs.poetry.withPlugins (
    ps: [
      ps.poetry-plugin-export
    ]
  );
```

**Comment by sandydoo** [comment]
I can't tell if `withPlugins` actually hooks these up properly 🤔 Would be curious to see how this all works out, specifying the feed, etc.

**Comment by ianepreston** [comment]
Cool! As a start I'll test that out. If it works then I think everything on the Devenv side is good to go and I just have to figure out how to get `artifacts-keyring` published as a poetry plugin. I've got to do some day job stuff for the next bit but I'll report back on this when I've had a chance to test. Thanks again for your help.

**Comment by ianepreston** [comment]
Ok, took me a little longer to get to this than I'd hoped. Here's what I found:

`poetry-plugin-export` is not a good test because the `poetry export` command works without it. Maybe the functionality got added into the main app. `poetry-audit-plugin` is a good test candidate though because `poetry audit` doesn't run on a fresh environment. A one line change in my `devenv.nix` from `package = pkgs.poetry;` to `package = pkgs.poetry.withPlugins (ps: [ ps.poetry-audit-plugin ]);` was sufficient to make `poetry audit` run. So that's pretty sweet! I now have to figure out how to create a poetry plugin for artifacts keyring, but looks like we're all good on the devenv side. Thanks for your help!
